File: python/astal/binding.py
import re


def kebabify(string):
    return re.sub(r"([a-z])([A-Z])", r"\1-\2", string).replace("_", "-").lower()


class Binding:
    def __init__(self, emitter, prop=None):
        self.emitter = emitter
        self.prop = kebabify(prop) if prop else None
        self.transform_fn = lambda v: v

    def __str__(self):
        return f"Binding<{self.emitter}{', ' + self.prop if self.prop else ''}>"

    def as_(self, fn):
        bind = Binding(self.emitter, self.prop)
        bind.transform_fn = lambda v: fn(self.transform_fn(v))
        return bind

    def get(self):
        if hasattr(self.emitter, "get") and callable(self.emitter.get):
            return self.transform_fn(self.emitter.get())

        return self.transform_fn(self.emitter[f"get_{self.prop}"]())

    def subscribe(self, callback):
        if hasattr(self.emitter, "subscribe") and callable(self.emitter.subscribe):
            return self.emitter.subscribe(lambda _: callback(self.get()))

        i = self.emitter.connect(f"notify::{self.prop}", lambda: callback(self.get()))
        return lambda: self.emitter.disconnect(i)

File: lua/astal/binding.lua
local lgi = require("lgi")
local GObject = lgi.require("GObject", "2.0")

---@class Binding
---@field emitter object
---@field property? string
---@field transformFn function
local Binding = {}

---@param emitter object
---@param property? string
---@return Binding
function Binding.new(emitter, property)
    return setmetatable({
        emitter = emitter,
        property = property,
        transformFn = function(v)
            return v
        end,
    }, Binding)
end

function Binding:__tostring()
    local str = "Binding<" .. tostring(self:get())
    if self.property ~= nil then
        str = str .. ", " .. self.property
    end
    return str .. ">"
end

function Binding:get()
    if type(self.emitter.get) == "function" then
        return self.transformFn(self.emitter:get())
    end
    return self.transformFn(self.emitter[self.property])
end

---@param transform fun(value: any): any
---@return Binding
function Binding:as(transform)
    local b = Binding.new(self.emitter, self.property)
    b.transformFn = function(v)
        return transform(self.transformFn(v))
    end
    return b
end

---@param callback fun(value: any)
---@return function
function Binding:subscribe(callback)
    if type(self.emitter.subscribe) == "function" then
        return self.emitter:subscribe(function()
            callback(self:get())
        end)
    end
    local id = self.emitter.on_notify:connect(function()
        callback(self:get())
    end, self.property, false)
    return function()
        GObject.signal_handler_disconnect(self.emitter, id)
    end
end

Binding.__index = Binding
return Binding
